File: 2week/근우/homework/5.employee-importance.py
"""
# Definition for Employee.
"""
from typing import List
from collections import deque


class Employee:
    def __init__(self, id: int, importance: int, subordinates: List[int]):
        self.id = id
        self.importance = importance
        self.subordinates = subordinates


class Solution:
    def getImportance(self, employees: List["Employee"], id: int) -> int:
        employee_dict = {}
        for employee in employees:
            employee_dict[employee.id] = employee
        ans = 0
        que = deque({id})
        while que:
            em = employee_dict[que.popleft()]
            ans += em.importance
            for s in em.subordinates:
                que.append(s)
        return ans

File: 2week/남규/5. Employee Importance.cpp
/*
// Definition for Employee.
class Employee {
public:
    int id;
    int importance;
    vector<int> subordinates;
};
*/

class Solution {
public:
    int getImportance(vector<Employee*> employees, int id) {
        stack<int> stk;
        stk.push(id);
        int ans = 0;
        
        // Time Complexity: O(N)
        // Space Complexity: O(N)
        unordered_map<int, Employee*> idToEmp;
        
        for(auto employee : employees){
            idToEmp[employee->id] = employee;
        }
        
        // unordered_set<int> visited;
        
        while(!stk.empty()){
            int curr = stk.top();
            ans += idToEmp[curr]->importance;
            stk.pop();
            
            for(int i = 0; i < idToEmp[curr]->subordinates.size(); i++){
                int next = idToEmp[curr]->subordinates[i];
                /*if(visited.find(next) != visited.end()) continue;
                visited.insert(next);*/
                stk.push(next);
            }
        }
        return ans;
    }
};
